proxy/grpcproxy/watcher_single.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2016 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcproxy
+
+import (
+	"github.com/coreos/etcd/clientv3"
+	"golang.org/x/net/context"
+)
+
+type watcherSingle struct {
+	// ch delievers events received from the etcd server
+	ch clientv3.WatchChan
+
+	// sws is the stream this watcherSingle attached to
+	sws *serverWatchStream
+
+	w watcher
+
+	rev int64 // current revision
+
+	// cancel is used to cancel the watch request sent to etcd server
+	cancel context.CancelFunc
+	stopc  chan struct{}
+}
+
+func newWatcherSingle(wch clientv3.WatchChan, c context.CancelFunc, w watcher, sws *serverWatchStream) *watcherSingle {
+	return &watcherSingle{
+		sws: sws,
+
+		ch: wch,
+		w:  w,
+
+		cancel: c,
+		stopc:  make(chan struct{}),
+	}
+}
+
+func (ws watcherSingle) run() {
+	for {
+		select {
+		case wr, ok := <-ws.ch:
+			if !ok {
+				return
+			}
+			ws.rev = wr.Header.Revision
+			ws.w.send(wr)
+
+			// a watcherSingle can promote itself to a watchergroup if it already caught up
+			// with the current revision.
+			canPromote := wr.Header.Revision == wr.Events[len(wr.Events)-1].Kv.ModRevision
```

Comment:
passing around `canPromote` is kind of clumsy, maybe keep the last observed modRevision in the `watcherSingle` struct and have `func (ws *watcherSingle) canPromote() bool { return ws.rev == ws.lastModRev }`?